oximeter/db/src/client/oxql.rs

```diff
@@ -356,25 +371,96 @@ impl Client {
             "coalesced" => ?&preds,
         );
 
-        // We generally run 2 SQL queries for each OxQL query:
+        // We generally run a few SQL queries for each OxQL query:
         //
-        // 1. Fetch all the fields consistent with the query.
-        // 2. Fetch the consistent samples.
-        let mut query_summaries = Vec::with_capacity(2);
-        let all_fields_query =
-            self.all_fields_query(&schema, preds.as_ref())?;
-        let (summary, info) = self
-            .select_matching_timeseries_info(&all_fields_query, &schema)
-            .await?;
-        debug!(
-            query_log,
-            "fetched information for matching timeseries keys";
-            "n_keys" => info.len(),
-        );
-        query_summaries.push(summary);
+        // - Some number of queries to fetch the timeseries keys that are
+        // consistent with it.
+        // - Fetch the consistent samples.
+        //
+        // Note that there are often 2 or more queries needed for the first
+        // case. In particular, there is one query required for each independent
+        // time range in the query (including when a time range isn't
+        // specified).
+        //
+        // For example, consider the filter operation:
+        //
+        // ```
+        // filter some_predicate || (timestamp > @now() - 1m && other_predicate)
+        // ```
+        //
+        // That is, we return all timepoints for things where `some_predicate`
+        // is true, and only the last minute for those satisfying
+        // `other_predicate`. If we simply drop the timestamp filter, and run
+        // the two predicates conjoined, we would erroneously return only the
+        // last minute for everything, including those satisfying
+        // `some_predicate`.
+        //
+        // So instead, we need to run one query for each of those, fetch the
+        // keys associated with it, and then independently select the
+        // measurements satisfying both the time range and key-consistency
+        // constraints. Thankfully that can be done in one query, albeit a
+        // complicated one.
+        //
+        // Convert any outer predicates to DNF, and split into disjoint key
+        // groups for the measurement queries.
+        let disjoint_predicates = if let Some(preds) = preds.as_ref() {
+            let simplified = preds.simplify_to_dnf()?;
+            debug!(
```

Comment:
Yeah, that sounds really useful. Right now, we're making a child logger specific to the OxQL query, so we should be able to filter the output with something like `looker` to just those messages. It would be helpful to have a more complete representation of the "plan" here, since right now we're emitting things like individual disjunctions that we operate on.

I'll keep it in mind as we add more planning / optimization to this, thanks!